### ISSUE TITLE
Fetched diagram cannot be located if image output directory does not match imagesdir

### ISSUE THREAD

**Issue by mojavelinux** [Mogztter/asciidoctor-kroki#373]
If the image output directory does not match the value of `imagesdir`, such as when `imagesoutdir` is specified, the converter cannot locate the fetched diagram. This particularly affects Asciidoctor PDF, which must read the image in order to embed it.

The current workaround for this situation in Asciidoctor Diagram is to use an absolute path as the image target if the `data-uri` attribute is set on the document. See:

* https://github.com/asciidoctor/asciidoctor-pdf/blob/main/lib/asciidoctor/pdf/converter.rb#L139-L140
* https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/diagram_processor.rb#L238

Since setting the `data-uri` has other implications, we may decide to settle on a different attribute name.

Since we plan to add support for the `imagesdir` on the image node in core (see https://github.com/asciidoctor/asciidoctor/issues/3660), one approach is to rely on that enhancement and set the `imagesdir` attribute on the generated node if it differs from the normal location. That way, the converter doesn't have to do any extra work.

**Comment by mojavelinux** [comment]
Related tests in Asciidoctor PDF: https://github.com/asciidoctor/asciidoctor-pdf/blob/main/spec/diagram_spec.rb#L38-L76

**Comment by Mogztter** [comment]
@mojavelinux should we anticipate this change and set the `imagesdir` attribute on the generated node today?

> Since we plan to add support for the imagesdir on the image node in core (see https://github.com/asciidoctor/asciidoctor/issues/3660), one approach is to rely on that enhancement and set the imagesdir attribute on the generated node if it differs from the normal location. That way, the converter doesn't have to do any extra work.

So, if `imagesoutdir` is specified we should set the `imagesdir` attribute on the generated node? Maybe we should always set the `imagesdir` on the generated node?

**Comment by mojavelinux** [comment]
Yes, I'm confident that we can add the check for imagesdir on the node in 2.1. I'll revive https://github.com/asciidoctor/asciidoctor/pull/3661

> Maybe we should always set the imagesdir on the generated node?

Can't hurt, esp since this is essentially an internal negotiation with the converter.

**Comment by mojavelinux** [comment]
I think this needs to be applied to the Ruby gem as well since the behavior currently differs from Asciidoctor Diagram.

**Comment by ggrossetie** [comment]
> I think this needs to be applied to the Ruby gem as well since the behavior currently differs from Asciidoctor Diagram.

Good point 👍🏻 
But as far as I know, https://github.com/asciidoctor/asciidoctor/pull/3661 is not yet released right?

**Comment by mojavelinux** [comment]
Correct. Though Asciidoctor Diagram manages to make it work without that change.

**Comment by ggrossetie** [comment]
Actually it was implemented in both!

**Comment by mojavelinux** [comment]
I believe what Asciidoctor Diagram does is account for the delta from imagesdir to imagesoutdir when computing the target of the generated image. That way, the image will be resolved by the converter no matter what imagesdir is set to.

**Comment by ggrossetie** [comment]
Verified this against Asciidoctor Diagram implementation and did some testing.

Asciidoctor Diagram `diagram_processor.rb` computes the delta between the general output dir and the actual image output dir, then sets it as an `imagesdir` override on the generated image node (rather than the document). Asciidoctor Diagram gates this behind an experimental opt-in attribute (`diagram-autoimagesdir`, #110). It's not the default there.

asciidoctor-kroki already implements the same node-level `imagesdir` override in both extensions (`src/fetch.js`, `ruby/.../extension.rb`), applied unconditionally whenever `kroki-fetch-diagram` writes a file. No opt-in needed, since it's a harmless no-op when the underlying Asciidoctor core doesn't support node-level `imagesdir` yet.

Will keep an eye on the `v2.0.x` branch for the backport.

**Comment by mojavelinux** [comment]
I'm confident this works when using Asciidoctor Diagram against the released version of Asciidoctor without any assignment to the imagesdir on the node. I have tested it with the Antora PDF extension and that's how I know it behaves properly.

**Comment by mojavelinux** [comment]
It's crucial that this be fixed in Asciidoctor Kroki without an Asciidoctor release or else Asciidoctor Kroki cannot be successfully used with the Antora PDF extension. It ends up looking in the wrong location for the generated image.

**Comment by mojavelinux** [comment]
Okay, there's another piece to this. Asciidoctor Diagram seems to recognize when the path can be absolute and writes it that way. When you convert to HTML without data-uri, then Asciidoctor Diagram does not get the path correct.

Try this with Asciidoctor Diagram and see that it works:

```
asciidoctor-pdf -a "diagram-cachedir=$PWD/.cache/asciidoctor-diagram" -a imagesoutdir=$PWD/docs/_images -a imagesdir -r asciidoctor-diagram docs/test.adoc
```

It also works for HTML if data-uri is set:

```
asciidoctor -a data-uri -a "diagram-cachedir=$PWD/.cache/asciidoctor-diagram" -a imagesoutdir=$PWD/docs/_images -a imagesdir -r asciidoctor-diagram docs/test.adoc
```

The imagesoutdir does not have to match imagesdir and support for the imagesdir on the node does not have to be available, but only if the image file is being read from disk.

**Comment by mojavelinux** [comment]
> Will keep an eye on the v2.0.x branch for the backport.

This change was not planned for 2.0.x. However, we may have to reconsider that.

**Comment by ggrossetie** [comment]
Dug into this further and reproduced it end-to-end with the real gems, and I think I now have the actual root cause and a fix that doesn't need any Asciidoctor release.

Asciidoctor PDF forces `data-uri` truthy on the document at converter init (`converter.rb:139`, "enabling data-uri forces Asciidoctor Diagram to produce absolute image paths").

The image is correctly written to `imagesoutdir`, and the node-level `imagesdir` override is correctly set, but Asciidoctor PDF ignores it (as expected, since the released core doesn't support it yet) and looks in the document's plain `imagesdir` instead.

Asciidoctor Diagram doesn't hit this because when `data-uri` is truthy, `diagram_processor.rb` sets `target` to the diagram's *absolute file path* instead of a relative name + `imagesdir` override. An absolute path bypasses `imagesdir`/`imagesoutdir` resolution entirely, so it works on any released Asciidoctor.

Asciidoctor Kroki `create_image_src` never looks at `data-uri` at all! It only branches on `kroki-fetch-diagram`, always writes to `imagesoutdir`, and always relies on the (currently inert) `imagesdir` node override. So under Asciidoctor PDF, it hits the exact failure above.

I believe that Asciidoctor PDF can consume a `data:` URI as the image `target` without any file on disk at all. It can, cleanly, no warnings, render images correctly. And it turns out the JavaScript/Node.js extension already does exactly this.

So the fix in the Ruby extension could be: when `data-uri` is set, fetch the diagram and return a `data:` URI as `target` directly (no disk write, no `imagesdir` dependency at all). Bringing Ruby in line with what JS already does, and with Diagram's outcome (works without any Asciidoctor imagesdir-on-node support) but via a simpler mechanism.

Wanted to check this reasoning with you before implementing, since it changes the fix from "wait on the imagesdir-on-node override to ship" to "handle data-uri explicitly in Ruby, today." 😄

**Comment by mojavelinux** [comment]
> Asciidoctor PDF forces data-uri truthy on the document at converter init (converter.rb:139, "enabling data-uri forces Asciidoctor Diagram to produce absolute image paths").

Aha! That's right. You finally triggered my memory of how we decided to do this, and why we do it that way in Asciidoctor PDF.

> I believe that Asciidoctor PDF can consume a data: URI as the image target without any file on disk at all. 

That's correct.

> Wanted to check this reasoning with you before implementing, since it changes the fix from "wait on the imagesdir-on-node override to ship" to "handle data-uri explicitly in Ruby, today." 😄

After giving this more thought, I realize that the sum of the assumptions I was going off of aren't going to work, so a change may be warranted in Antora PDF anyway.

With that said, I do think that keeping in alignment with Asciidoctor Diagram (until imagesdir is supported on the node) makes the most sense.

I'm going to follow-up on the Antora part of this in the chat.

**Comment by ggrossetie** [comment]
> With that said, I do think that keeping in alignment with Asciidoctor Diagram (until imagesdir is supported on the node) makes the most sense.

I'd lean the other way, for two reasons.

First, internal consistency: the JS/Node.js extension already treats `data-uri` as "embed the diagram inline". It fetches the content and returns a `data:` URI as `target`, no file written. The Ruby extension currently does neither that nor Diagram's absolute-path trick. It just ignores `data-uri` entirely. So today Ruby and JS Kroki already disagree with each other. Aligning Ruby with Diagram would still leave that JS/Ruby split in place, just moved to a different axis.

Second, and more importantly, using an absolute-path when `data-uri` is defined conflates two things that aren't actually related. `data-uri` is documented as "the user wants this image embedded inline." Asciidoctor PDF repurposes that same attribute purely as a signal to Asciidoctor Diagram to sidestep `imagesdir` resolution ("enabling data-uri forces Asciidoctor Diagram to produce absolute image paths", its own comment), even though PDF isn't asking for base64 embedding at all, it's asking for a resolvable path. That coupling only exists because Asciidoctor Diagram implementation happens to check that one attribute for a different purpose. It's not a designed contract, it's a hack of convenience.

If Kroki instead takes `data-uri` at face value, fetch and return an actual `data:` URI, no disk write, no dependency on `imagesdir`/`imagesoutdir` at all, it gets the same practical outcome for Antora PDF, while actually doing what `data-uri` says on the tin, and without waiting on imagesdir-on-node support. Same result, but for the right reason, and it unifies Ruby with what JS already does.

Happy to implement it this way unless you see a downside I'm missing.

**Comment by mojavelinux** [comment]
I take your point. It's clear now that the functionality we have in Asciidoctor Diagram was to work around a limitation in Asciidoctor and Asciidoctor PDF, yet it's not the best long-term design. The right approach is to use the imagesdir on the node, so we'll have to reevaluate the decision to not backport that.

This change is less urgent now that I've reverted the custom imageoutdir in Antora PDF. We now have a way to use Asciidoctor Kroki and Asciidoctor Diagram. Although not ideal, there were other problems to solve anyway.

So let's stick with what you implemented since it is better in the long term.
